pennylane_forest/qpu.py

```diff
@@ -67,11 +93,154 @@ def __init__(self, device, *, shots=1024, active_reset=False, load_qc=True, **kw
         aspen_match = re.match(r"Aspen-\d+-([\d]+)Q", device)
         num_wires = int(aspen_match.groups()[0])
 
-        super(QVMDevice, self).__init__(
-            num_wires, shots, **kwargs
-        )  # pylint: disable=bad-super-call
+        super().__init__(num_wires, shots, **kwargs)
 
         if load_qc:
             self.qc = get_qc(device, as_qvm=False, connection=self.connection)
+        else:
+            self.qc = get_qc(device, as_qvm=True, connection=self.connection)
 
         self.active_reset = active_reset
+        self.symmetrize_readout = symmetrize_readout
+        self.calibrate_readout = calibrate_readout
+
+    def pre_measure(self):
+        # pass
+        ## code below borrowed from `qvm.py`
+        """Run the QVM"""
+        # pylint: disable=attribute-defined-outside-init
+        for e in self.obs_queue:
+            wires = e.wires
+
+            if e.name in ["PauliX", "PauliY", "PauliZ", "Identity", "Hadamard"]:
+                pass
+
+            elif e.name == "Hermitian":
+                # For arbitrary Hermitian matrix H, let U be the unitary matrix
+                # that diagonalises it, and w_i be the eigenvalues.
+                H = e.parameters[0]
+                Hkey = tuple(H.flatten().tolist())
+
+                if Hkey in self._eigs:
+                    # retrieve eigenvectors
+                    U = self._eigs[Hkey]["eigvec"]
+                else:
+                    # store the eigenvalues corresponding to H
+                    # in a dictionary, so that they do not need to
+                    # be calculated later
+                    w, U = np.linalg.eigh(H)
+                    self._eigs[Hkey] = {"eigval": w, "eigvec": U}
+
+                # Perform a change of basis before measuring by applying U^ to the circuit
+                self.apply("QubitUnitary", wires, [U.conj().T])
+
+        prag = Program(Pragma("INITIAL_REWIRING", ['"PARTIAL"']))
+
+        if self.active_reset:
+            prag += RESET()
+
+        self.prog = prag + self.prog
+
+        qubits = list(self.prog.get_qubits())
+        ro = self.prog.declare("ro", "BIT", len(qubits))
+        for i, q in enumerate(qubits):
+            self.prog.inst(MEASURE(q, ro[i]))
+
+        self.prog.wrap_in_numshots_loop(self.shots)
+
+        if "pyqvm" in self.qc.name:
+            bitstring_array = self.qc.run(self.prog)
+        else:
+            executable = self.qc.compile(self.prog)
+            bitstring_array = self.qc.run(executable=executable)
+
+        self.state = {}
+        for i, q in enumerate(qubits):
+            self.state[q] = bitstring_array[:, i]
```

Comment:
Yeah, good call. I wasn't too happy about this code duplication either. I've made `QPUDevice` a subclass of `QVMDevice` again, and gotten rid of `probabilities`. I can't completely get rid of `pre_measure` though because there's a small difference here, specifically in the instruction `if e.name in ["PauliX", "PauliY", "PauliZ", "Identity", "Hadamard"]: pass`. Maybe we could refactor some of the `pre_measure` code out, and over-write only the relevant parts?

Reply:
I've re-factored some code out of `pre_measure` into a separate `pre_rotations` to avoid duplication.